components/omega/src/ocn/OceanState.cpp

```diff
@@ -0,0 +1,385 @@
+                                                      NormalVelocity[CurLevel]);
+   Err = OMEGA::IOField::attachData<OMEGA::Array2DR8>("LayerThickness",
+                                                      LayerThickness[CurLevel]);
+
+} // end defineIOFields
+
+//------------------------------------------------------------------------------
+// Destroy parallel decompositions
+void OceanState::finalizeParallelIO() {
+
+   int Err = 0; // default return code
+
+   // Destroy the IO decomp for arrays with (NCells) dimensions
+   Err = IO::destroyDecomp(CellDecompR8);
+   if (Err != 0)
+      LOG_CRITICAL("OceanState: error destroying cell IO decomposition");
+
+   // Destroy the IO decomp for arrays with (NEdges) dimensions
+   Err = IO::destroyDecomp(EdgeDecompR8);
+   if (Err != 0)
+      LOG_CRITICAL("OceanState: error destroying edge IO decomposition");
+
+} // end finalizeParallelIO
+
+//------------------------------------------------------------------------------
+// Read Ocean State
+void OceanState::read() {
+
+   I4 Err;
+
+   // Read LayerThickness
+   int LayerThicknessID;
+   Err = IO::readArray(LayerThicknessH[0].data(), NCellsAll, "layerThickness",
+                       StateFileID, CellDecompR8, LayerThicknessID);
+   if (Err != 0)
+      LOG_CRITICAL("OceanState: error reading layerThickness");
+
+   // Read NormalVelocity
+   int NormalVelocityID;
+   Err = IO::readArray(NormalVelocityH[0].data(), NEdgesAll, "normalVelocity",
+                       StateFileID, EdgeDecompR8, NormalVelocityID);
+   if (Err != 0)
+      LOG_CRITICAL("OceanState: error reading normalVelocity");
+
+} // end read
+
+//------------------------------------------------------------------------------
+// Perform copy to device for state variables
+void OceanState::copyToDevice(int TimeLevel) {
+
+   deepCopy(LayerThickness[TimeLevel], LayerThicknessH[TimeLevel]);
+   deepCopy(NormalVelocity[TimeLevel], NormalVelocityH[TimeLevel]);
+
+} // end copyToDevice
+
+//------------------------------------------------------------------------------
+// Perform copy to host for state variables
+void OceanState::copyToHost(int TimeLevel) {
+
+   deepCopy(LayerThicknessH[TimeLevel], LayerThickness[TimeLevel]);
+   deepCopy(NormalVelocityH[TimeLevel], NormalVelocity[TimeLevel]);
+
+} // end copyToHost
+
+//------------------------------------------------------------------------------
+// Perform time level update
+void OceanState::updateTimeLevels() {
+
+   int NewLevel = NTimeLevels - 1;
+
+   // Update time levels for layer thickness
+   copyToHost(NewLevel);
+   MeshHalo->exchangeFullArrayHalo(LayerThicknessH[NewLevel], OMEGA::OnCell);
+   copyToDevice(NewLevel);
+
+   Array2DR8 Temp;
+   HostArray2DR8 TempH;
+
+   for (int Level = 0; Level < NTimeLevels - 1; Level++) {
+      Temp                      = LayerThickness[Level + 1];
+      LayerThickness[Level + 1] = LayerThickness[Level];
+      LayerThickness[Level]     = Temp;
```

Comment:
Here and below, using`std::swap` from `<algorithm>` might be slightly more readable.
```suggestion
      std::swap(LayerThickness[Level + 1], LayerThickness[Level]);

```